libraries/Microsoft.Bot.Builder/CloudAdapterBase.cs

```diff
@@ -0,0 +1,320 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Security.Principal;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Connector;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Microsoft.Bot.Builder
+{
+    /// <summary>
+    /// An adapter that implements the Bot Framework Protocol and can be hosted in different cloud environmens both public and private.
+    /// </summary>
+    public class CloudAdapterBase : BotAdapter
+    {
+        internal const string InvokeResponseKey = "BotFrameworkAdapter.InvokeResponse";
+
+        private ICloudEnvironment _cloudEnvironment;
+        private HttpClient _httpClient;
+        private ILogger _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CloudAdapterBase"/> class.
+        /// </summary>
+        /// <param name="cloudEnvironment">The cloud environment used for validating and creating tokens.</param>
+        /// <param name="httpClient">The IHttpClientFactory implementation this adapter should use.</param>
+        /// <param name="logger">The ILogger implementation this adapter should use.</param>
+        protected CloudAdapterBase(
+            ICloudEnvironment cloudEnvironment,
+            HttpClient httpClient = null,
+            ILogger logger = null)
+        {
+            _cloudEnvironment = cloudEnvironment;
```

Comment:
This one is not optional and we don't seem to have a default, so probably we want to validate for null here

```suggestion
            _cloudEnvironment = cloudEnvironment ?? throw new ArgumentNullException(nameof(cloudEnvironment));
```